README.md
# Recent Global Posts Feed

**INACTIVE NOTICE: This plugin is unsupported by WPMUDEV, we've published it here for those technical types who might want to fork and maintain it for their needs.**

## Recent Global Posts feed takes advantage of Post Indexer to add an RSS feed of all the posts on your entire Multisite or BuddyPress network - simply, efficiently and quickly.

##### This powerful plugin allows you to:

*   Creates a **Post feed URL for your entire network** which anyone can subscribe to
*   Provides a simple **Global Post Feed widget**

### Setup is easy

Once installed your new feed is created at:

<pre>http://yoursite.com/feed/globalpostsfeed</pre>

You can even access custom post type feeds by visiting:

<pre>http://yoursite.com/feed/globalpostsfeed?posttype=customposttype</pre>

Just substitute 'customposttype' with your preferred post type.

### Bundled widget for easy sharing

A built-in widget makes global feeds easy to share. Best of all you can add a simple '_Subscribe to our Network post feed'_ by adding the Recent Global Post feed widget.

### To Install:

1\. Install the Post Indexer

*   The Post Indexer is designed to index posts on your network and needs to be installed for the Recent Global Posts Feed plugin to work.

2. Install Recent Global Posts Feed.

3. Once installed your new feed is created at:

<pre>http://yoursite.com/feed/globalpostsfeed</pre>

*   You can access feeds for custom post types by using http://yoursite.com/feed/globalpostsfeed?posttype=customposttype, substituting customposttype with your custom post type
*   You can check if it is working correctly by posting some new posts and confirming they appear in your global feed.

4\.  Install Recent Global Posts Widget

*   Visit **Network Admin -> Plugins** to Network Activate the Recent Global Posts Widget
*   Your Recent Global Posts Widget is added to **Appearance > Widgets** of your main site.
*   By default, the widget is available only on the main site of your network. This behavior can be changed by following the instructions below.
*   The Recent Global Posts Feed is for public posts only. Posts from privates sites don't appear in the feed.

### Enabling Widget for all sites

By default the Recent Global Posts Widget is only enabled for use by the main site. You can enable it for all sites on your network as follows:

1.  Open your **wp-config.php** file using a text editing program like [EditPlus](http://www.editplus.com/) 

2.  Add the following constant to that file just before the line that says "That's all, stop editing" `define( 'RECENT_GLOBAL_POSTS_FEED_WIDGET_MAIN_BLOG_ONLY', false );` 

3.  Save and re-upload your wp-config.php file. Note: if you ever want to limit the use of the widget to the main site only again, simply change "false" to "true" in the code you just added.

### To Use

1.  Go to **Appearance > Widgets**

2.  Add the Recent Global Posts Widget to your sidebar. 

3.  Check out the configuration options 


### Known Issues

If you are having difficulty getting your feed to validate, please go to Settings > Permalinks in your admin and click Save Changes. That will re-save your settings and updated the rewrite rules. To check if your feed validates properly, visit [http://validator.w3.org/feed/](http://validator.w3.org/feed/)


changelog.txt
Plugin Name: Recent Global Posts Feed
Author: Andrew Billits

Change Log:
----------------------------------------------------------------------
----------------------------------------------------------------------
2.1 - 05/10/2011
----------------------------------------------------------------------
- Converted to widget class
- Added post type parameter

2.0 - 01/08/2011
----------------------------------------------------------------------
- Update for version 3.1

1.0.6 - 12/23/2010
----------------------------------------------------------------------
- Fixed: Widget Dragging Problem

1.0.5 - 07/13/2010
----------------------------------------------------------------------
- Fixed occassional 404 header responses causing invalid feed

1.0.4 - 10/13/2009
----------------------------------------------------------------------
- Switched user login to display name

1.0.3 - 08/06/2009
----------------------------------------------------------------------
- Added better compatibility with feed readers

1.0.2 - 06/19/2009
----------------------------------------------------------------------
- Added widget

1.0.1 - 05/17/2009
----------------------------------------------------------------------
- Fixed a feed bug resulting in an error whne using IE
- Modified rss tags for better compatibility with feed readers

1.0.0 - 04/13/2009
----------------------------------------------------------------------
- Initial Release.


widget-recent-global-posts-feed.php
<?php
/*
Plugin Name: Recent Posts Feed Widget
Description:
Author: Andrew Billits (Incsub)
Version: 2.0
Author URI:
*/

/*
Copyright 2007-2009 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/* -------------------- Update Notifications Notice -------------------- */
if ( !function_exists( 'wdp_un_check' ) ) {
  add_action( 'admin_notices', 'wdp_un_check', 5 );
  add_action( 'network_admin_notices', 'wdp_un_check', 5 );
  function wdp_un_check() {
    if ( !class_exists( 'WPMUDEV_Update_Notifications' ) && current_user_can( 'edit_users' ) )
      echo '<div class="error fade"><p>' . __('Please install the latest version of <a href="http://premium.wpmudev.org/project/update-notifications/" title="Download Now &raquo;">our free Update Notifications plugin</a> which helps you stay up-to-date with the most stable, secure versions of WPMU DEV themes and plugins. <a href="http://premium.wpmudev.org/wpmu-dev/update-notifications-plugin-information/">More information &raquo;</a>', 'wpmudev') . '</a></p></div>';
  }
}
/* --------------------------------------------------------------------- */

//------------------------------------------------------------------------//
//---Config---------------------------------------------------------------//
//------------------------------------------------------------------------//
$recent_global_posts_feed_widget_main_blog_only = 'yes'; //Either 'yes' or 'no'
//------------------------------------------------------------------------//
//---Hook-----------------------------------------------------------------//
//------------------------------------------------------------------------//

//------------------------------------------------------------------------//
//---Functions------------------------------------------------------------//
//------------------------------------------------------------------------//
function widget_recent_global_posts_feed_init() {
	global $wpdb, $recent_global_posts_feed_widget_main_blog_only;

	// Check for the required API functions
	if ( !function_exists('register_sidebar_widget') || !function_exists('register_widget_control') )
		return;

	// This saves options and prints the widget's config form.
	function widget_recent_global_posts_feed_control() {
		global $wpdb;
		$options = $newoptions = get_option('widget_recent_global_posts_feed');
		if ( $_POST['recent-global-posts-feed-submit'] ) {
			$newoptions['recent-global-posts-feed-title'] = $_POST['recent-global-posts-feed-title'];
			$newoptions['recent-global-posts-feed-rss-image'] = $_POST['recent-global-posts-feed-rss-image'];
		}
		if ( $options != $newoptions ) {
			$options = $newoptions;
			update_option('widget_recent_global_posts_feed', $options);
		}
		if ( empty( $options['recent-global-posts-feed-title'] ) ) {
			$options['recent-global-posts-feed-title'] = __('Site Posts Feed');
		}
		if ( empty( $options['recent-global-posts-feed-rss-image'] ) ) {
			$options['recent-global-posts-feed-rss-image'] = 'show';
		}
	?>
				<div style="text-align:left">

				<label for="recent-global-posts-feed-title" style="line-height:35px;display:block;"><?php _e('Title', 'widgets'); ?>:<br />
                <input class="widefat" id="recent-global-posts-feed-title" name="recent-global-posts-feed-title" value="<?php echo $options['recent-global-posts-feed-title']; ?>" type="text" style="width:95%;" />
                </label>
				<label for="recent-global-posts-feed-rss-image" style="line-height:35px;display:block;"><?php _e('RSS Image', 'widgets'); ?>:<br />
                <select name="recent-global-posts-feed-rss-image" id="recent-global-posts-feed-rss-image" style="width:95%;">
                <option value="show" <?php if ($options['recent-global-posts-feed-rss-image'] == 'show'){ echo 'selected="selected"'; } ?> ><?php _e('Show'); ?></option>
                <option value="hide" <?php if ($options['recent-global-posts-feed-rss-image'] == 'hide'){ echo 'selected="selected"'; } ?> ><?php _e('Hide'); ?></option>
                </select>
                </label>
				<input type="hidden" name="recent-global-posts-feed-submit" id="recent-global-posts-feed-submit" value="1" />
				</div>
	<?php
	}
// This prints the widget
	function widget_recent_global_posts_feed($args) {
		global $wpdb, $current_site;
		extract($args);
		$defaults = array('count' => 10, 'username' => 'wordpress');
		$options = (array) get_option('widget_recent_global_posts_feed');

		foreach ( $defaults as $key => $value )
			if ( !isset($options[$key]) )
				$options[$key] = $defaults[$key];

		?>
		<?php echo $before_widget; ?>
			<?php
			if ( $options['recent-global-posts-feed-rss-image'] == 'hide' ) {
	            echo $before_title . '<a href="http://' . $current_site->domain . $current_site->path . 'wp-content/recent-global-posts-feed.php" >' . __($options['recent-global-posts-feed-title']) . '</a>' . $after_title;
			} else {
	            echo $before_title . '<a href="http://' . $current_site->domain . $current_site->path . 'wp-content/recent-global-posts-feed.php" ><img src="http://' . $current_site->domain . $current_site->path . 'wp-includes/images/rss.png" /> ' . __($options['recent-global-posts-feed-title']) . '</a>' . $after_title;
			}
			?>
		<?php echo $after_widget; ?>
<?php
	}
	// Tell Dynamic Sidebar about our new widget and its control
	if ( $recent_global_posts_feed_widget_main_blog_only == 'yes' ) {
		if ( $wpdb->blogid == 1 ) {
			register_sidebar_widget(array(__('Recent Global Posts Feed'), 'widgets'), 'widget_recent_global_posts_feed');
			register_widget_control(array(__('Recent Global Posts Feed'), 'widgets'), 'widget_recent_global_posts_feed_control');
		}
	} else {
		register_sidebar_widget(array(__('Recent Global Posts Feed'), 'widgets'), 'widget_recent_global_posts_feed');
		register_widget_control(array(__('Recent Global Posts Feed'), 'widgets'), 'widget_recent_global_posts_feed_control');
	}
}

add_action('widgets_init', 'widget_recent_global_posts_feed_init');

?>


recent-global-posts-feed.php
<?php
/*
Plugin Name: Recent Global Posts Feed
Plugin URI: http://premium.wpmudev.org/project/recent-global-posts-feed
Description: A sitewide feed of all the latest posts from all the blogs across your entire site
Author: Andrew Billits (Incsub)
Version: 2.1
WDP ID: 70
*/

/*
Copyright 2007-2009 Incsub (http://incsub.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (Version 2 - GPLv2) as published by
the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

require_once('../wp-load.php');

/* -------------------- Update Notifications Notice -------------------- */
if ( !function_exists( 'wdp_un_check' ) ) {
  add_action( 'admin_notices', 'wdp_un_check', 5 );
  add_action( 'network_admin_notices', 'wdp_un_check', 5 );
  function wdp_un_check() {
    if ( !class_exists( 'WPMUDEV_Update_Notifications' ) && current_user_can( 'edit_users' ) )
      echo '<div class="error fade"><p>' . __('Please install the latest version of <a href="http://premium.wpmudev.org/project/update-notifications/" title="Download Now &raquo;">our free Update Notifications plugin</a> which helps you stay up-to-date with the most stable, secure versions of WPMU DEV themes and plugins. <a href="http://premium.wpmudev.org/wpmu-dev/update-notifications-plugin-information/">More information &raquo;</a>', 'wpmudev') . '</a></p></div>';
  }
}
/* --------------------------------------------------------------------- */

//------------------------------------------------------------------------//
//---Config---------------------------------------------------------------//
//------------------------------------------------------------------------//
$number = $_GET['number'];
if ( empty( $number ) ) {
	$number = '25';
}

$posttype = $_GET['posttype'];
if(empty($posttype)) {
	$posttype = 'post';
}
//------------------------------------------------------------------------//
//---Functions------------------------------------------------------------//
//------------------------------------------------------------------------//
global $wpdb;

$query = $wpdb->prepare( "SELECT * FROM " . $wpdb->base_prefix . "site_posts WHERE site_id = %d AND blog_public = '1' AND post_type = %s ORDER BY post_published_gmt DESC LIMIT %d", $current_site->id, $posttype, $number);
$posts = $wpdb->get_results( $query, ARRAY_A );

if ( count( $posts ) > 0 ) {
	$last_published_post_date_time = $wpdb->get_var( $wpdb->prepare( "SELECT post_published_gmt FROM " . $wpdb->base_prefix . "site_posts WHERE site_id = %d AND blog_public = '1' AND post_type = %s ORDER BY post_published_gmt DESC LIMIT 1", $current_site->id, $posttype ) );
}

header('Content-Type: text/xml; charset=' . get_option('blog_charset'), true);
$more = 1;

?>
<?php echo '<?xml version="1.0" encoding="'.get_option('blog_charset').'"?'.'>'; ?>
<rss version="2.0"
	xmlns:content="http://purl.org/rss/1.0/modules/content/"
	xmlns:wfw="http://wellformedweb.org/CommentAPI/"
	xmlns:dc="http://purl.org/dc/elements/1.1/"
	xmlns:atom="http://www.w3.org/2005/Atom"
	xmlns:sy="http://purl.org/rss/1.0/modules/syndication/"
>

<channel>
	<title><?php bloginfo_rss('name'); ?> <?php _e('Posts'); ?></title>
	<atom:link href="<?php self_link(); ?>" rel="self" type="application/rss+xml" />
	<link><?php bloginfo_rss('url') ?></link>
	<description><?php bloginfo_rss("description") ?></description>
	<pubDate><?php echo mysql2date('D, d M Y H:i:s +0000', $last_published_post_date_time, false); ?></pubDate>
	<?php the_generator( 'rss2' ); ?>
	<language><?php echo get_option('rss_language'); ?></language>
    <?php
	//--------------------------------------------------------------------//
	if ( count( $posts ) > 0 ) {
		foreach ($posts as $post) {
			$author_display_name = $wpdb->get_var("SELECT display_name FROM " . $wpdb->base_prefix . "users WHERE ID = '" . $post['post_author'] . "'");
			?>
			<item>
				<title><?php echo strip_tags(stripslashes($post['post_title'])); ?></title>
				<link><?php echo $post['post_permalink']; ?></link>
				<comments><?php echo $post['post_permalink'] . '#comments'; ?></comments>
				<pubDate><?php echo mysql2date('D, d M Y H:i:s +0000', $post['post_published_gmt'], false); ?></pubDate>
				<dc:creator><?php echo $author_display_name; ?></dc:creator>
				<guid isPermaLink="false"><?php echo $post['post_permalink']; ?></guid>
                <description><![CDATA[<?php echo wp_trim_excerpt($post['post_content']); ?>]]></description>
                <content:encoded><![CDATA[<?php echo $post['post_content']; ?>]]></content:encoded>
				<wfw:commentRss><?php echo $post['post_permalink'] . 'feed/'; ?></wfw:commentRss>
			</item>
			<?php
		}
	}
	//--------------------------------------------------------------------//
	?>
</channel>
</rss>
<?php
//------------------------------------------------------------------------//
?>
